### PR TITLE
v2podman add container init

### PR DESCRIPTION
**Pull request by baude**
add the ability to init a container both local and remote

Signed-off-by: Brent Baude <bbaude@redhat.com>

### CODE REVIEW

File: pkg/api/handlers/libpod/containers.go [hunk 1]
```diff
@@ -285,3 +285,23 @@ func Restore(w http.ResponseWriter, r *http.Request) {
...
+	name := utils.GetName(r)
+	runtime := r.Context().Value("runtime").(*libpod.Runtime)
+	ctr, err := runtime.LookupContainer(name)
+	if err != nil {
+		utils.ContainerNotFound(w, name, err)
+		return
+	}
+	err = ctr.Init(r.Context())
+	if errors.Cause(err) == define.ErrCtrStateInvalid {
+		utils.Error(w, "container already initd", http.StatusNotModified, err)
```
Comment (jwhonce) [inline pkg/api/handlers/libpod/containers.go:299]:
```suggestion
		utils.Error(w, "container already init'ed", http.StatusNotModified, err)
```
Seems better english.

File: pkg/api/server/register_containers.go [hunk 2]
```diff
@@ -1411,6 +1410,29 @@ func (s *APIServer) registerContainersHandlers(r *mux.Router) error {
 	r.HandleFunc(VersionedPath("/containers/{name}/changes"), s.APIHandler(compat.Changes))
 	r.HandleFunc("/containers/{name}/changes", s.APIHandler(compat.Changes))
 	r.HandleFunc(VersionedPath("/libpod/containers/{name}/changes"), s.APIHandler(compat.Changes))
-
+	// swagger:operation POST /libpod/containers/{name}/init libpod libpodInitContainer
+	// ---
+	// tags:
+	//  - containers
+	// summary: Initialize a container
+	// description: Performs all tasks necessary for starting the container but does not start the container.
```
Comment (jwhonce) [inline pkg/api/server/register_containers.go:1418]:
```suggestion
	// description: Performs all tasks necessary for initializing the container but does not start the container.
```

File: pkg/domain/infra/abi/containers.go [hunk 3]
```diff
@@ -794,3 +794,17 @@ func (ic *ContainerEngine) ContainerCleanup(ctx context.Context, namesOrIds []st
 	}
 	return reports, nil
 }
+
+func (ic *ContainerEngine) ContainerInit(ctx context.Context, namesOrIds []string, options entities.ContainerInitOptions) ([]*entities.ContainerInitReport, error) {
+	var reports []*entities.ContainerInitReport
+	ctrs, err := getContainersByContext(options.All, options.Latest, namesOrIds, ic.Libpod)
```
Comment (jwhonce) [inline pkg/domain/infra/abi/containers.go:800]:
```suggestion
	ctrs, err := getContainersByContext(options.All, options.Latest, namesOrIds, ic.Libpod)
```
Just querying the containers causes them to be initialized?  I suspect a C&P error here.  I am also confused how you are going to create the container if a "not found" error is going to return.

File: pkg/domain/infra/tunnel/containers.go [hunk 4]
```diff
@@ -338,3 +338,19 @@ func (ic *ContainerEngine) ContainerDiff(ctx context.Context, nameOrId string, _
...
 }
+
+func (ic *ContainerEngine) ContainerInit(ctx context.Context, namesOrIds []string, options entities.ContainerInitOptions) ([]*entities.ContainerInitReport, error) {
+	var reports []*entities.ContainerInitReport
+	ctrs, err := getContainersByContext(ic.ClientCxt, options.All, namesOrIds)
+	if err != nil {
+		return nil, err
+	}
+	for _, ctr := range ctrs {
+		err := containers.ContainerInit(ic.ClientCxt, ctr.ID)
```
Comment (jwhonce) [inline pkg/domain/infra/tunnel/containers.go:349]:
```suggestion
		err := containers.ContainerInit(ic.ClientCxt, ctr.ID)
```
This looks better.

File: pkg/domain/infra/tunnel/containers.go [hunk 5]
```diff
@@ -338,3 +338,19 @@ func (ic *ContainerEngine) ContainerDiff(ctx context.Context, nameOrId string, _
 func (ic *ContainerEngine) ContainerCleanup(ctx context.Context, namesOrIds []string, options entities.ContainerCleanupOptions) ([]*entities.ContainerCleanupReport, error) {
 	return nil, errors.New("not implemented")
 }
+
+func (ic *ContainerEngine) ContainerInit(ctx context.Context, namesOrIds []string, options entities.ContainerInitOptions) ([]*entities.ContainerInitReport, error) {
+	var reports []*entities.ContainerInitReport
+	ctrs, err := getContainersByContext(ic.ClientCxt, options.All, namesOrIds)
+	if err != nil {
```
Comment (jwhonce) [inline pkg/domain/infra/tunnel/containers.go:345]: Isn't container not found an error?

File: pkg/domain/infra/abi/containers.go (hunk 3 above)
Comment (baude) [reply pkg/domain/infra/abi/containers.go:800]: This is correct.  If the container input is bad (i.e. not found), then it returns and does not proceed.  This is the exact behavor of v1.

File: pkg/domain/infra/tunnel/containers.go (hunk 5 above)
Comment (baude) [reply pkg/domain/infra/tunnel/containers.go:345]: this behavior is same as v1, if input is bad, return before attempting any inits.

**Pull request merged**